245 hfp optimization

danieldunderfelt opened the pull request:
HFP investigation, optimization and fixes related to tickets https://github.com/HSLdevcom/bultti/issues/245, https://github.com/HSLdevcom/bultti/issues/241, https://github.com/HSLdevcom/bultti/issues/221 and https://github.com/HSLdevcom/bultti/issues/217.

UI PR: https://github.com/HSLdevcom/bultti-ui/pull/70

Main changes:

- Checks which stops to load for each route, loads only stops relevant to Bultti.
- Updates and optimisations for features that use HFP data
- Refactors and updates to HFP functions

Divinare commented on src/hfp/hfpUtils.ts:
```diff
@@ -135,7 +69,53 @@ export function createSpecificJourneyKey(
   )}_${item.unique_vehicle_id}`
 }
+
+export async function createExistingEventsChecker(
```
createExistingStopEventsChecker?

danieldunderfelt replied:
No, this just checks that the event isn't already in the database.

And I've removed it completely in the 217 branch :D

Divinare commented on src/hfp/hfpUtils.ts:
```diff
@@ -135,7 +69,53 @@ export function createSpecificJourneyKey(
+
+  for (let uuidChunk of existingUuidChunks) {
+    let set = new Set(uuidChunk)
+    existingEventUuids.push(set)
```
Interesting, making a set for each million row just in case to prevent memory issues or smth

danieldunderfelt replied:
Yes! The Set only supports a certain amount of entries, so I had to split them up into multiple sets.

Divinare commented on src/inspection/inspectionLinks.ts:
```diff
@@ -200,3 +200,31 @@ export async function checkIfLinkedInspectionsAreUpdated(inspection: Inspection)
+
+// Check from which other inspections this inspection is linked.
+// eg. if arg is a pre-inspection, returns all post-inspections where it is used.
+export async function getLinkedFromInspections(inspection: Inspection): Promise<Inspection[]> {
```
Nit:
```suggestion
export async function getLinkedInspections(inspection: Inspection): Promise<Inspection[]> {
```

danieldunderfelt replied:
This is the reverse of "linked inspections", as this function finds where the `inspection` is linked _from_, hence the name. If you have a better suggestion I will consider it, but "from" is an important word in the name here.

Divinare replied:
Ah this one is tough to name!

I came up with:
```getInspectionsLinkedFromInspection()
...
let inspectionsLinkedFromInspection = await repo.find({
```

Divinare commented on src/inspection/inspectionLinks.ts:
```diff
@@ -200,3 +200,31 @@ export async function checkIfLinkedInspectionsAreUpdated(inspection: Inspection)
+
+  let repo = getRepository(Inspection)
+
+  let linkedFromInspections = await repo.find({
```
Nit: could it be like this?
```suggestion
  let linkedInspections = await getRepository(Inspection).find({
```

danieldunderfelt replied:
Sure, we can inline the repo creation. But as I said previously, "from" is an important distinction here and it tries to tell you what's going on.

Divinare reviewed (changes requested):
Left some minor comments, otherwise LGTM.

Hfp-loading related code start to get clear! That observedDeparturesTask is still confusing, a lot of things were refactored / changed there, this task got a bit too wide I guess.

danieldunderfelt commented on src/inspection/InspectionResolver.ts:
```diff
@@ -446,29 +450,40 @@ export class InspectionResolver implements ResolverInterface<Inspection> {
     }
 
+    inspection = inspectionRepo.merge(inspection, assignInspectionInput)
+    inspection.versionStackIdentifier = createStackIdentifier(inspection)
```
The idea here is to extract the `inspectionDate` stuff from the input object and fetch the inspectionDate with the inspectionId. Then, what's left of the inspectionInput is put into the `assignInspectionInput` var. Then, outside the `if`, the whole `assignInspectionInput` is written to the Inspection entity. For PRE-inspections, this would include everything from the input.

For non-PRE, we still need to assign the rest of the inspectionInput to the entity, there are `name`, `startDate` and `endDate` props that would be ignored if we removed the `merge` row.

I suppose the `InspectionInput` props can be assigned one-by-one, but the current code works so I don't really see a reason to change it.

danieldunderfelt replied:
I simplified the code a bit anyway!

Divinare replied:
Yeah now I think it makes more sense!

I misunderstood the code because of this line:
```
    if (inspection.status !== InspectionStatus.Draft) {
      useInspectionInput = pick(inspectionInput, 'name')
    }
```
Thought useInspectionInput would have only 'name' but yea it's only when inspection.status !== InspectionStatus.Draft... Should have read more carefully!

danieldunderfelt commented on src/queue/queueManager.ts:
```diff
@@ -4,6 +4,7 @@ import prexit from 'prexit'
 import { getTopicTaskCallbacks, topicTasks } from './topicTasks'
 import { workerTasks } from './workerTasks'
 import { isJobStalled } from './queueUtils'
+import { DISABLE_CACHE } from '../constants'
```
Dang, I try to remove these with the webstorm feature but some slip through.

danieldunderfelt commented on src/inspection/inspectionReportsPreCache.ts:
```diff
@@ -25,24 +26,30 @@ export async function runInspectionReportsPreCache({
   shouldReloadReport?: boolean
   reportNames?: string[]
 }): Promise<void> {
+  if (DISABLE_PRE_CACHE) {
```
Yes!

danieldunderfelt replied:
It seems to already be there... maybe I added it yesterday.

danieldunderfelt commented on src/hfp/hfpStorage.ts:
```diff
@@ -24,64 +23,92 @@ export async function getHfpBlobs(date: string) {
+          eventTypes.includes(hfpRow.event_type || '') &&
+          hfpRow.uuid &&
+          !checkEventUuidExists(hfpRow.uuid) &&
+          (await eventFilter(hfpRow))
```
No, that's not the point. The eventFilter is instantiated outside the loop, because it sets up a cache map that remembers the result of queries. If it would be instantiated here, there would be no caching since it would be recreated on each loop iteration.

I don't really like passing it down either but it is quite necessary.

Divinare replied:
Maybe could have a few comments:
```
// Returns a callback that returns a boolean whether given hfpRow is a stop event or not
// Uses cache so this has to created only once
export function createHfpStopsFilter(weekStartDate: Date) {
```

danieldunderfelt commented on src/hfp/hfpStopsFilter.ts:
```diff
@@ -0,0 +1,33 @@
+  const filterCache = new Map<string, boolean>()
+
+  return (hfpEvent: HfpRow): Promise<boolean> => {
+    if (!hfpEvent.stop || !hfpEvent.route_id || !hfpEvent.direction_id) {
```
Yes, this is more of a safeguard. Theoretically there could be an object coming it where all fields are `null`, and we don't want those.

danieldunderfelt commented on src/hfp/hfpStopsFilter.ts:
```diff
@@ -0,0 +1,33 @@
+      .then(stops => stops.map(stop => stop.stop_id))
+      .then(routeStops => {
+        let result =
+          routeStops.length === 0 ? false : routeStops.some(stopId => hfpEvent.stop === stopId)
```
Of course it's a filter! It checks that the HFP event is from a stop that is relevant to Bultti and filters out the ones that are not.

`HfpStopsFilter` is also quite apt, since it filters the HFP rows based on the stop.

I really don't get this comment.

danieldunderfelt commented on .env.production:
```diff
@@ -45,12 +45,13 @@ BULTTI_SCHEMA=bultti
 
-HFP_TASK_DATE_CONCURRENCY=4
-WORKER_CONCURRENCY=3
+WORKER_CONCURRENCY=1
```
This config determines how many tasks one worker does concurrently. There are more than one worker yes.

This is a conservative value, and it is just a default. It can also be set in the Gitlab deploy repo. If we monitor the workers and see that they can do multiple tasks at once the number can be increased.

And yes, I also don't want to leave configs that are unused!

danieldunderfelt commented on src/hfp/hfpStorage.ts:
```diff
@@ -24,64 +23,92 @@ export async function getHfpBlobs(date: string) {
 > {
+  let checkEventUuidExists = await createExistingEventsChecker(date)
+
   let container = await getContainer(hfpContainerName)
```
Well, the Azure library calls it a container...

danieldunderfelt commented on src/hfp/hfpStorage.ts:
```diff
@@ -24,64 +23,92 @@ export async function getHfpBlobs(date: string) {
+          let journeyId = journeys.get(journeyKey)
+
+          if (!journeyId) {
+            createdJourney = HfpJourney.createFromHfpRow(hfpRow, blobName)
```
Yes, this could be very useful to do! Although it won't tell the whole story about missing events, it may tell a part of it.

Card => https://github.com/HSLdevcom/bultti/issues/249

danieldunderfelt commented:
I did minor refactors and cleaned up some unused env vars!

About the `observedDeparturesTask`, I refactored it into multiple smaller functions to make it more understandable. There are not really any actual _changes_, just moving code around.

Divinare reviewed (approved):
Maybe could have a few comments:
```
// Returns a callback that returns a boolean whether given hfpRow is a stop event or not
// Uses cache so this has to created only once
export function createHfpStopsFilter(weekStartDate: Date) {
```

Divinare commented:
Hfp-loading seems to work!
